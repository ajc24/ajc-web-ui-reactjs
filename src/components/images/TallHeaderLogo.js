/**
 * Developed by Anthony Cox in 2025
 */
import React from 'react';
import PropTypes from 'prop-types';
import DecorativeImage from './DecorativeImage';
import './css/image.css';

/* Set the size dimensions for header logos (if rendered) */
const tallHeaderLogoDimensions = {
  height: {
    solo: {
      noUpperBorder: 304,
      withUpperBorder: 300,
    },
  },
  width: {
    default: 328,
  },
};

/**
 * Tall Header Logo component which describes an image which is used solely for decoration in the header of
 * a web application. This component is intended for use with the Tall Header component.
 */
class TallHeaderLogo extends React.Component {
  /**
   * Initialise the tall header logo component
   * @param {any} props 
   */
  constructor(props) {
    super(props);
    this.state = {
      height: tallHeaderLogoDimensions.height.solo.noUpperBorder,
      width: tallHeaderLogoDimensions.width.default,
    }
    this.handleScreenWidth = this.handleScreenWidth.bind(this);
    this.setHeight = this.setHeight.bind(this);
  }

  componentDidMount() {
    if (this.props.isTopBorderDisplayed === true) {
      this.setHeight(tallHeaderLogoDimensions.height.solo.withUpperBorder);
    }
    /* Initial set of the header logo size based on the screens current size */
    this.handleScreenWidth();

    /* Watch over all future window resize events - we will want to alter the height of the logo depending on how the title text and subtitle text elements are rendered */
    window.addEventListener('resize', this.handleScreenWidth);
  }

  handleScreenWidth() {
    if (this.props.headerTitleTextId !== undefined) {
      /* Get the title text heading element and determine its height */
      const titleTextElement = document.querySelector(`div[id="${this.props.headerTitleTextId}--title-text"] > h1`);
      if (titleTextElement !== null) {
        const titleTextElementHeight = titleTextElement.getBoundingClientRect().height;

        if (this.props.subtitleTextId === undefined) {
          /* No subtitle text is rendered - base the height of the logo on the height of the title text alone */
          let newLogoHeight = tallHeaderLogoDimensions.height.solo.noUpperBorder;
          if (this.props.isTopBorderDisplayed === true) {
            newLogoHeight = tallHeaderLogoDimensions.height.solo.withUpperBorder;
          }
          this.setHeight(newLogoHeight - titleTextElementHeight);
        }
      }
    }
  }

  /**
   * Sets the height of the logo
   * @param {number} newHeight 
   */
  setHeight(newHeight) {
    this.setState({
      height: newHeight,
    });
  }

  render() {
    const containerCss = 'image-tall-header-logo-alignment';
    return (
      <div className={containerCss}>
        <DecorativeImage id={`${this.props.id}--tall-header-logo`} src={this.props.src} height={this.state.height} width={this.state.width} />
      </div>
    );
  }
}
TallHeaderLogo.propTypes = {
  /** The unique identifier which was given to the header title text component. */
  headerTitleTextId: PropTypes.string,
  /** The unique identifier for this component. */
  id: PropTypes.string.isRequired,
  /* Switch to indicate whether the Small Header components top border is displayed or not */
  isTopBorderDisplayed: PropTypes.bool,
  /** The image data to be displayed. */
  src: PropTypes.oneOfType([ PropTypes.string, PropTypes.object ]).isRequired,
  /** The unique identifier which was given to the subtitle text component. */
  subtitleTextId: PropTypes.string,
};
export default TallHeaderLogo;
