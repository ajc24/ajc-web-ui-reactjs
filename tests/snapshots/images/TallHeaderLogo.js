/**
 * Developed by Anthony Cox in 2024
 */
import 'jsdom-global/register';
import React from 'react';
import { TestDev } from 'ajc-testing-code';
import TallHeaderLogo from '../../../src/components/images/TallHeaderLogo';
import storybookLogo from '../../../stories/images/files/storybook-logo-long-tall.svg';

describe('Tall Header Logo', () => {
  describe('Default logo image with no header upper border rendered', () => {
    let snapshot;
    
    beforeAll(() => {
      snapshot = TestDev.createSnapshot(
        <React.Fragment>
          <TallHeaderLogo id="tall-header-logo" src={storybookLogo} />
        </React.Fragment>
      );
    });

    it('verifies the snapshot for the component', () => {
      expect(snapshot).toMatchSnapshot();
    });
  });

  describe('Tall header logo image with no header upper border rendered', () => {
    let snapshot;
    
    beforeAll(() => {
      snapshot = TestDev.createSnapshot(
        <React.Fragment>
          <TallHeaderLogo id="tall-header-logo" isTopBorderDisplayed={false} src={storybookLogo} />
        </React.Fragment>
      );
    });

    it('verifies the snapshot for the component', () => {
      expect(snapshot).toMatchSnapshot();
    });
  });

  describe('Tall header logo image with header upper border rendered', () => {
    let snapshot;
    
    beforeAll(() => {
      snapshot = TestDev.createSnapshot(
        <React.Fragment>
          <TallHeaderLogo id="tall-header-logo" isTopBorderDisplayed={true} src={storybookLogo} />
        </React.Fragment>
      );
    });

    it('verifies the snapshot for the component', () => {
      expect(snapshot).toMatchSnapshot();
    });
  });
});
