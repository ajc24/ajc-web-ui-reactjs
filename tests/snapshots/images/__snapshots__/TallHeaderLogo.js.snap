// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tall Header Logo Default logo image with no header upper border rendered verifies the snapshot for the component 1`] = `"<div class="image-tall-header-logo-alignment"><img id="tall-header-logo--tall-header-logo--decorative-image" class="image" width="328" height="304" alt="" src="[object Object]"></div>"`;

exports[`Tall Header Logo Tall header logo image with header upper border rendered verifies the snapshot for the component 1`] = `"<div class="image-tall-header-logo-alignment"><img id="tall-header-logo--tall-header-logo--decorative-image" class="image" width="328" height="300" alt="" src="[object Object]"></div>"`;

exports[`Tall Header Logo Tall header logo image with no header upper border rendered verifies the snapshot for the component 1`] = `"<div class="image-tall-header-logo-alignment"><img id="tall-header-logo--tall-header-logo--decorative-image" class="image" width="328" height="304" alt="" src="[object Object]"></div>"`;
